libia2/threads.c

```diff
@@ -0,0 +1,52 @@
+#include <pthread.h>
+
+#include "ia2.h"
+
+void init_stacks(void);
+void protect_tls(void);
+
+struct ia2_thread_thunk {
+  void *(*fn)(void *);
+  void *data;
+};
+
+void *ia2_thread_begin(void *arg) {
+  struct ia2_thread_thunk *thunk = (struct ia2_thread_thunk *)arg;
+  void *(*fn)(void *) = thunk->fn;
+  void *data = thunk->data;
+  /* Free the thunk. */
+  munmap(arg, sizeof(struct ia2_thread_thunk));
+
+  init_stacks();
+  protect_tls();
+
+  /* TODO: switch to compartment stack before calling `fn` */
+
+  return fn(data);
+}
+
+int __real_pthread_create(pthread_t *restrict thread,
+                          const pthread_attr_t *restrict attr, void *(*fn)(),
+                          void *data);
+
+int __wrap_pthread_create(pthread_t *restrict thread,
+                          const pthread_attr_t *restrict attr, void *(*fn)(),
+                          void *data) {
+  /* Allocate a thunk for the thread to call `ia2_thread_begin` before the
+  provided thread body function. */
+  void *mmap_res = mmap(NULL, sizeof(struct ia2_thread_thunk),
```

Comment:
I'm guessing you don't want to use malloc here, but noting why would be nice.

Reply:
Done.